### DISCUSSION TITLE
Multivariate time series forecast example

### DISCUSSION THREAD

**Discussion by Dreamlone** [winedarksea/AutoTS#129] [Q&A]
Hi! I could not find an example that allows running the library for multivariate time series forecasting. However, the documentation states that the library supports this functionality 

Please give an example of how the library can be used 1) to fit multivariate time series forecasting model and 2) to use the model on new data, the same as for [94](https://github.com/winedarksea/AutoTS/discussions/94)

Data example:
| datetime   | target ts | ts exog |
| :--------: | :-----------: | :---: |
| 01-01-2022 | 0           | 10   |
| 02-01-2022 | 1           | 11   |

![image](https://user-images.githubusercontent.com/52418437/175960795-a96ec336-25eb-4335-ba20-8f80b7ef2ee0.png)

target ts - the time series I want to predict. For the target time series and for the exogenous one, only the historical values are known

By the way, I do this for a [pytsbe](https://github.com/ITMO-NSS-team/pytsbe) repository for benchmarking time series forecasting algorithms. If you want to participate in this open-source project for comparison (and implement methods from AutoTS, for example, so that the AutoTS can predict as good as possible), you are welcome :)

**Comment by winedarksea** [comment]
Every example given in the documentation is multivariate. The way to make the model focus on only one target instead of all is to pass series `weights` to `.fit()`
You can see an example here: https://github.com/winedarksea/AutoTS/blob/master/extended_tutorial.md#another-example where traffic_volume is the target in this example.
```python
weights_hourly = {'traffic_volume': 20}

...

model = model.fit(
    df_wide,
    weights=weights_hourly,
)
```
🇺🇦

**Reply by Dreamlone** [reply]
Thanks so much for the quick reply - I'll make a try!

**Reply by adanebehr** [reply]
In this example you give, the wide dataframe with multiple columns is passed directly to the model.fit() call. But in other examples and in the documentation I see either create_regressor() or fake_regressor() functions used and then their results passed to model.fit() and model.predict() calls. What is the difference between these methodologies? Also, on that same train of thought, what is the difference between the create_regressor() and fake_regressor() functions?

**Reply by winedarksea** [reply]
@adanebehr  Fake regressor just generates random numbers. Completely useless in real world situations, but useful for testing.

The model expects a `df` as the primary input to `.fit()`. This is your endogenous data and are the time series you want to predict. Sometimes you have other data to pass in as an additional detail, an exogenous variable sometimes called `time_varying_known` (things you know in advance) which is passed as the `future_regressor`. And what `create_regressor` does is construct a *future* regressor from an existing dataset of `time_varying_unknown` (stuff you don't know about for the future) by lagging it as well as adding some `time_varying_known` information from date_parts and holidays. To be clear, `create_regressor` and any sort of `future_regressor` is completely optional, and the main thing to focus on is getting your primary input dataframe of wide or long data in correctly.

Terminology is definitely confusing coming from outside of the time series world, where you only ever have one type of X input. Here the input data is both X and Y, with additional X's from a different time period as optional. Hope that isn't too confusing!

**Reply by adanebehr** [reply]
thanks for always sending such thorough and informative responses. I feel like I've learned a lot just working with your package. Just to regurgitate my understanding of your response and make sure I'm following:
So it sounds like just passing the wide dataframe to .fit() does something similar to create_regressor in the background, but create_regressor gives you something you can use to scenario model (by modifying values of the forecasted regressor that is the second variable returned by create_regressor)  and incorporates some standard known future info, like holidays. And the values of the regressors are "lagged", meaning the future_regressor is just the regressor shifted into the future by the length of the forecast?
And I guess the numbers given by create_regressor represent some sort of correlative relationship between the columns in the dataframe? 

To jump to the chase in why I'm asking all these questions and trying to understand:
I have two time series. Business cash flows and federal funds rate. The business cash flows are what I want to predict, and I hypothesize that the fed funds rate impacts the rate of business growth. Because the cash flows are the target, I've used weights like you suggest to remove importance of predicting fed funds rate in selecting the model. BUT Cash flows are non-stationary, wheras fed funds rate is stationary.... Is there a problem with this? Or can I just let the magic black box do its thing? It seems like the fed funds rate needs to be modeled/forecasted with a separate model. For example, I would like to constrain the fed funds rate predictions to be within the min and max of the dataset, but I dont want to constrain the cashflows because they trend upwards.

**Reply by winedarksea** [reply]
What it is doing with the dataframe passed really depends on the model it is passing it to. Probably the most like you are imagining is the [rolling_x_regressor](https://github.com/winedarksea/AutoTS/blob/master/autots/models/sklearn.py#L25) used in the MultivariateRegression.

Horizontal ensembling is used by default. That means each series uses separately the best model available for it. Mosaic ensembling is a more extreme variation with each timestep a separate model. Printing the model details will tell you (in json) what model was chosen for each input series.

Finally it sounds like you may benefit from using the `constraint` argument.  You can use an arbitrarily high constraint for the first series and a more relevant one for the second.
```python
{
    "constraint_method": "absolute",
    "constraint_regularization": 1,
    "upper_constraint": pd.Series([99999999, 10],
    "lower_constraint": pd.Series([-9999999, -10]),
    "bounds": True,
}
```

**Reply by adanebehr** [reply]
That is very cool. I'm pleased that it is working. I had ensembles turned off before, so I see now that now the multiple series are modeled separately.

Is there a way to check if there is some dependency between the two time series being modeled? I.e. rather than the two being independently forecasted, with no impact to one another.

**Reply by winedarksea** [reply]
Look at the models chosen and see if there are among those in the model table that are listed as multivariate. That would generally suggest they are finding useful information from the other series.

There are of course other ways, correlation, Granger casuality, the causalimpacts package, and using statsmodels to model with statistical tests to make an argument  for whether they have an impact, but in terms of this package, the simple way is to see if a multivariate model outperformed the univariate models.
